Use `chunked_vector` in `append_entries_request`

mmaslankaprv opened the pull request:
`model::record_batch_reader` was not the best abstraction to represent
the batches in `append_entries_request`. Number of batches contained in
the request is known upfront and they are always materialized in memory.
Therefore it is more convenient and easier to store batches as a plain
`chunked_vector` instead of the reader.
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
r/types: do not use reader in append entries request

`model::record_batch_reader` was not the best abstraction to represent
the batches in `append_entries_request`. Number of batches contained in
the request is known upfront and they are always materialized in memory.
Therefore it is more convenient and easier to store batches as a plain
`chunked_vector` instead of the reader.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

dotnwat reviewed (commented):
deja vu did we just review a similar pr?

mmaslankaprv commented:
kind of similar, this one is a follow up, getting rid of the reader abstraction from RPC request

mmaslankaprv committed:
r/fixture: use serde wrapper for append entries request

In order to match the real life scenario changed raft fixture to use
serde wrapper for append entries request as wrapper is used in normal
operations.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--94c5062f-687e-4c54-9190-6ca667f21744-->
<!--TEST RESULTS-->
<!--b3ca6a09-426c-4b19-8ce4-cefea49a901f-->
<!--TEST RESULTS-->
<!--713e753a-b947-4900-ac4a-63011d65e6e3-->
<!--TEST RESULTS-->
<!--c8bf6813-f904-4ccb-aee5-616f4bf89090-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60953">build#60953</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60953#019483cf-5edb-44c4-b103-4d61503275f7 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-173a-4b08-912a-c4ea07cbc523 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_creating_when_cluster_misconfigured2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-173a-4b08-912a-c4ea07cbc523 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.None.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-1738-4738-9ab7-ab595829935e | FLAKY | 1/3 |
| rptest.tests.e2e_iam_role_test.ShortLivedCredentialsTests.test_short_lived_credentials | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-173a-4b08-912a-c4ea07cbc523 | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndHydrationTimeoutTest.test_hydration_completes_on_timeout | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-1738-4738-9ab7-ab595829935e | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTest.test_reset_from_cloud.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-1738-4738-9ab7-ab595829935e | FLAKY | 1/5 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTest.test_write.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-1738-4738-9ab7-ab595829935e | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_decommissioning_cancel_ongoing_movements | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-1738-4738-9ab7-ab595829935e | FLAKY | 1/2 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_partition_autobalancer_sanction.disable_license=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-173a-4b08-912a-c4ea07cbc523 | FLAKY | 1/2 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_partition_autobalancer_sanction.disable_license=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-1738-4738-9ab7-ab595829935e | FLAKY | 1/2 |
| rptest.tests.partition_force_reconfiguration_test.PartitionForceReconfigurationTest.test_basic_reconfiguration.acks=1.restart=True.controller_snapshots=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-173a-4b08-912a-c4ea07cbc523 | FLAKY | 1/2 |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move_node_down | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-173a-4b08-912a-c4ea07cbc523 | FLAKY | 1/2 |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move_x_core.replication_factor=1.unclean_abort=False.recovery=no_recovery.compacted=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-1738-4738-9ab7-ab595829935e | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.PartitionMovementTest.test_overlapping_changes.num_to_upgrade=2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-173a-4b08-912a-c4ea07cbc523 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_node_with_unavailable_node | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-173a-4b08-912a-c4ea07cbc523 | FLAKY | 1/10 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_on_demand_rebalancing.partition_count=20 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-173a-4b08-912a-c4ea07cbc523 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_topic_hot_spots | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-1738-4738-9ab7-ab595829935e | FLAKY | 1/16 |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-1738-4738-9ab7-ab595829935e | FLAKY | 1/11 |
| rptest.tests.shard_placement_test.ShardPlacementTest.test_node_join.disable_license=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-1738-4738-9ab7-ab595829935e | FLAKY | 1/5 |
| rptest.tests.topic_creation_test.TopicRecreateTest.test_topic_recreation_while_producing.workload=ACKS_1.cleanup_policy=delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/60953#01948414-173a-4b08-912a-c4ea07cbc523 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60965">build#60965</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.controller_log_limiting_test.ControllerLogLimitPartitionBalancerTests.test_partition_balancer_with_limits | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9825-4641-9134-890279770f34 | FLAKY | 1/13 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9823-432c-9fb6-7109029b692b | FLAKY | 1/2 |
| rptest.tests.data_transforms_test.DataTransformsTest.test_consume_off_end.offset=None | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9825-4641-9134-890279770f34 | FLAKY | 1/2 |
| rptest.tests.e2e_iam_role_test.STSRoleFetchTests.test_write | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9825-4641-9134-890279770f34 | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTest.test_reset_from_cloud.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9823-432c-9fb6-7109029b692b | FLAKY | 1/2 |
| rptest.tests.full_node_recovery_test.FullNodeRecoveryTest.test_node_recovery.recovery_type=full | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9825-4641-9134-890279770f34 | FLAKY | 1/2 |
| rptest.tests.idempotency_test.IdempotencySnapshotDelivery.test_recovery_after_snapshot_is_delivered | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9825-4641-9134-890279770f34 | FLAKY | 1/2 |
| rptest.tests.partition_force_reconfiguration_test.PartitionForceReconfigurationTest.test_basic_reconfiguration.acks=-1.restart=True.controller_snapshots=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9825-4641-9134-890279770f34 | FLAKY | 1/3 |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move.replication_factor=1.unclean_abort=True.recovery=no_recovery.compacted=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9823-432c-9fb6-7109029b692b | FLAKY | 1/2 |
| rptest.tests.raft_periodic_flush_test.PeriodicFlushWithRelaxedConsistencyTest.test_changing_periodic_flush_threshold | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9825-4641-9134-890279770f34 | FLAKY | 1/8 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_multiple_nodes_to_the_cluster.partition_count=20 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9825-4641-9134-890279770f34 | FLAKY | 1/5 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_topic_hot_spots | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9823-432c-9fb6-7109029b692b | FLAKY | 1/10 |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9823-432c-9fb6-7109029b692b | FLAKY | 1/2 |
| rptest.tests.shadow_indexing_compacted_topic_test.TSWithAlreadyCompactedTopic.test_initial_upload | ducktape | https://buildkite.com/redpanda/redpanda/builds/60965#0194853e-9825-4641-9134-890279770f34 | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60987">build#60987</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.datalake_e2e_test.DatalakeE2ETests.test_topic_lifecycle.cloud_storage_type=CloudStorageType.S3.filesystem_catalog_mode=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/60987#019488e9-59a5-40ed-87fb-7ee03ba4f877 | FLAKY | 1/2 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/60987#019488e9-59a5-4052-bde0-5d364528bc03 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61037">build#61037</a></summary>


| test_id                                                                                                                                | test_kind   | job_url                                                                                   | test_status   | passed   |
|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_chunk_read_path_test.CloudStorageChunkReadTest.test_read_when_segment_size_smaller_than_chunk_size          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61037#01948d9b-c123-4963-a751-bbc9898b394b | FLAKY         | 1/2      |
| rptest.tests.cluster_metadata_upload_test.ClusterMetadataUploadTest.test_uploads_after_restart.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61037#01948d9b-c123-4963-a751-bbc9898b394b | FLAKY         | 1/2      |
</details>
<!--/body-->

dotnwat commented:
linter problem

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60965-->
<!--c448af63-4933-42ac-9c17-e6d15f17c26d-->
<!--title-->
## Retry command for [Build#60965](https://buildkite.com/redpanda/redpanda/builds/60965)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::TSWithAlreadyCompactedTopic.test_initial_upload
```
<!--/body-->

mmaslankaprv committed:
r/append_entries: drop not required share mutex

Signed-off-by: Michał Maślanka <michal@redpanda.com>

ztlpn commented on src/v/raft/tests/raft_fixture.cc:
```diff
@@ -307,8 +309,11 @@ ss::future<result<vote_reply>> in_memory_test_protocol::vote(
   model::node_id id, append_entries_request req, rpc::client_opts opts) {
-    return dispatch<append_entries_request, append_entries_reply>(
-      id, std::move(req), std::move(opts));
+    ;
```
unneeded `;`?

ztlpn commented on src/v/raft/types.h:
```diff
@@ -247,12 +247,15 @@ struct append_entries_request
+        return _batches;
+    }
 
     static append_entries_request make_foreign(append_entries_request&& req);
```
do we still need this? looks this is basically just a move

ztlpn commented on src/v/raft/tests/configuration_serialization_test.cc:
```diff
@@ -140,20 +139,24 @@ SEASTAR_THREAD_TEST_CASE(test_config_extracting_reader) {
     cfg_2.set_version(raft::group_configuration::v_6);
-    auto cfg_batch_2 = raft::details::serialize_configuration_as_batches(cfg_2);
+    auto cfg_batch_2 = raft::details::serialize_configuration_as_batch(cfg_2);
+    ss::circular_buffer<model::record_batch> cfg_batches_2;
```
not a big deal, but should these be chunked vectors as well?

mmaslankaprv replied:
Not really as for now we have the `for_each_ref_extract_configuration` expecting reader as a parameter

ztlpn commented on src/v/raft/consensus.cc:
```diff
@@ -2061,34 +2055,22 @@ consensus::do_append_entries(append_entries_request&& r) {
+              = last_matched + model::offset(b.header().last_offset_delta + 1);
+            if (
+              last_batch_offset > last_log_offset
+              || get_term(last_batch_offset) != b.term()) {
```
It is a bit annoying that we are checking every batch. Can we just find a mismatch point, or alternatively add a check that there is only one mismatch point (i.e. if we are in the else branch then `batches` should be empty)?
